Fixes for DynamicBayesianNetwork's bugs

nihil21 opened the pull request:
### Issue number(s) that this pull request fixes
- Fixes #1334 #1338 #1339 #1331 

### List of changes to the codebase in this pull request
- In `DynamicBayesianNetwork.py`, I added a new class called `DynamicNode` which represents a dDBN's node. This is useful since before it used standard tuples, which created confusion.
- In `DynamicBayesianNetwork.py`, I modified `In `DynamicBayesianNetwork` class such that it internally converts the nodes passed as tuples into `DynamicNode` objects.
- In `Independencies.py` I modified the function `_return_list_if_str` into `_return_list_if_not_collection`: now it checks if the argument is a list/tuple/set, and if it isn't then it encapsulates the argument into a list; in this way, it becomes compatible with `DynamicNode` objects (cf. #1339).
- In `Independencies.py` I modified the `join()` calls such that it can print correctly non-str events (cf. #1339).
- In `DynamicBayesianNetwork.py`, I added the `active_trail_nodes()` method which converts arguments into `DynamicNode` type, if necessary, and then calls the respective super method (cf. #1334).
- In `DynamicBayesianNetwork.py`, I added the `get_markov_blanket()` method which converts arguments into `DynamicNode` type, if necessary, calls the respective super method, and then add to the Markov blanket the nodes from the unrolled DBN (cf. #1331).
- In `DAG.py`, I modified line 670 and replaced `blanket_nodes.remove(node)` with `blanket_nodes.discard(node)` (cf. #1331).

nihil21 committed:
Added DynamicNode type to represent nodes in DynamicBayesianNetwork instead of tuples

nihil21 committed:
Changed Independencies.py to take into account non-str node types (e.g. DynamicNode)

nihil21 committed:
Added active_trail_nodes() method to DynamicBayesianNetwork which converts tuples to DynamicNode objects and calls super method

nihil21 committed:
Adjusted get_markov_blanket() method to take into account the unrolled DBN

ankurankan commented:
@nihil21 The tests are failing because of formatting. You can run `black .` from the pgmpy dir to fix it.

nihil21 committed:
Reformatting

nihil21 commented:
Thanks! I wasn't aware of the `black` tool for reformatting.

nihil21 commented:
I've looked into the failed tests, and as far as I can tell the problem is that the DBN takes in input tuples and gives in output `DynamicNode` objects due to the internal conversion, while the tests expect to receive a tuple. For instance, in the test `test_add_multiple_edges`:

- the DBN receives `[(("a", 0), ("b", 0)), (("a", 0), ("a", 1)), (("b", 0), ("b", 1))]` and converts each tuple into a `DynamicNode`;
- in output it gives `OutEdgeView([(a_0, b_0), (a_0, a_1), (b_0, b_1), (a_1, b_1)])`, where `a_0`, `b_0`, etc. are `DynamicNode` objects.

nihil21 committed:
Added equality operator between DynamicNode and Sequence (list, tuple)

nihil21 committed:
Added equality operator between DynamicNode and Sequence (list, tuple)

ankurankan committed:
Fixes tests and adds to_tuple method to DynamicNode class

ankurankan commented:
@nihil21 Thanks a lot for the PR. This should fix a lot of issues. I have also added a commit to this PR fixing the failing tests. Could you maybe add some tests for the two new methods `get_markov_blanket` and `active_trail_nodes`? 

And just a reminder, make sure to pull your origin/dev before making any changes (as I pushed my commit to your dev) to avoid any conflicts.

nihil21 committed:
Adjusted bug in get_markov_blanket() method: now it returns also the parents of the children from the next time slice

nihil21 committed:
Added tests for get_markov_blanket() and active_trail_nodes()

nihil21 commented:
You're welcome! 

I have pulled from my origin/dev, but there are still conflicts. Apart from that, I've corrected get_markov_blanket() such that it returns also the children's parents, and added the two test.

ankurankan committed:
Fixes tests

ankurankan committed:
Black fixes

ankurankan commented:
@nihil21 Thanks for adding the tests. You were getting the merge conflict because pgmpy/pgmpy:dev had new commits, so you needed to either rebase or merge that into your fork. I have now rebased pgmpy/pgmpy:dev on nihil21/pgmpy:dev to fix the conflict, and also did a minor fix for one of the failing tests. The PR looks good to me now. 

Also, you will probably need to do a hard reset on your local dev branch because it would have diverged from your origin/dev (because of rebase and force push). Something like:
```
git fetch origin
git reset --hard origin/dev
```

nihil21 commented:
Perfect, thank you!
